### PR TITLE
gh-101819: Refactor _io in preparation for module isolation

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-101819

### CODE REVIEW

File: Modules/_io/_iomodule.h [hunk 1]
```diff
@@ -34,9 +34,13 @@ extern PyType_Spec winconsoleio_spec;
  * with args=NULL, and return a new reference.
  * BUT when args=Py_True is passed, they return a borrowed reference.
  */
-extern PyObject* _PyIOBase_check_readable(PyObject *self, PyObject *args);
-extern PyObject* _PyIOBase_check_writable(PyObject *self, PyObject *args);
-extern PyObject* _PyIOBase_check_seekable(PyObject *self, PyObject *args);
+struct _io_state;  // Forward decl.
```
Comment (vstinner) [inline Modules/_io/_iomodule.h]: Here, you can define the type: ``typedef struct _io_state _PyIO_State;`` to use ``_PyIO_State`` to use in these functions. Later, you can define the structure (members).

See ``Include/pytypedefs.h`` for the principle.

**Review by vstinner — APPROVED** [review]
LGTM, I just have a coding style suggestion for the header file (_iomodule.h).

File: Modules/_io/_iomodule.h (hunk 1 above)
Comment (erlend-aasland) [reply Modules/_io/_iomodule.h]: Yeah, that looks better. Thanks, I've applied that!

File: Modules/_io/_iomodule.h [hunk 2]
```diff
@@ -140,7 +144,7 @@ extern Py_off_t PyNumber_AsOff_t(PyObject *item, PyObject *err);
 
 extern PyModuleDef _PyIO_Module;
 
-typedef struct {
+typedef struct _io_state {
```
Comment (vstinner) [inline Modules/_io/_iomodule.h]: Please remove the typedef here to prevent compiler warnings/errors. clang dislikes this if I recall correctly.

Comment (erlend-aasland) [reply Modules/_io/_iomodule.h]: Removed in 7aa8da76a45c237f80e86a4bd2d0542e8426e395

**Review by vstinner — APPROVED** [review]
LGTM. The _iomodule.h typedef LGTM.

**Pull request merged**